File: packages/azalee/src/search/smart-search.ts
/**
 * Smart Search Service - Context-aware, multilingual search with caching
 *
 * Features:
 * - Multilingual search (FR/EN/JP)
 * - Context-aware prioritization based on page
 * - Fuzzy matching with "did you mean" suggestions
 * - Request cancellation and debouncing
 * - Result caching
 */

import { detectMatchedLanguage, findClosestMatch } from "./fuzzy-match";

export type SearchableType =
	| "basara"
	| "character"
	| "skill"
	| "item"
	| "tactic"
	| "keshin"
	| "soul"
	| "miximax"
	| "awakening"
	| "modechange"
	| "aura"
	| "team"
	| "article"
	| "patchnote"
	| "passive"
	| "tweet"
	| "re"
	| "cross"
	| "level5"
	| "topic"
	| "doc";

export interface GlobalSearchResult {
	id: string;
	name: string;
	type: SearchableType;
	score: number;
	data: Record<string, unknown> | null;
}

/** Enhanced search result with language detection and suggestions */
export interface SmartSearchResult extends GlobalSearchResult {
	/** Language that matched (FR/EN/JP) */
	matchedLanguage?: "FR" | "EN" | "JP";
	/** Alternative name in matched language */
	matchedName?: string;
	/** "Did you mean?" suggestion if no exact match */
	suggestion?: string;
}

/** Search context for prioritization */
export type SearchContext =
	| "chara"
	| "skill"
	| "team"
	| "aura"
	| "passive"
	| "tactic"
	| "global"
	| "item"
	| "news"
	| "patchnotes";

/** Search configuration */
export interface SmartSearchConfig {
	/** Context for result prioritization */
	context?: SearchContext;
	/** Maximum results to return */
	limit?: number;
	/** Enable fuzzy suggestions */
	enableSuggestions?: boolean;
	/** Minimum score threshold */
	minScore?: number;
}

/** Cache entry with TTL */
interface CacheEntry {
	results: SmartSearchResult[];
	timestamp: number;
}

/** Search result cache (5 minute TTL) */
const CACHE_TTL = 5 * 60 * 1000;
const searchCache = new Map<string, CacheEntry>();

/**
 * Clé de cache : elle doit couvrir TOUTE option qui change la sortie.
 *
 * `minScore` et `enableSuggestions` manquaient : deux appels de même requête et
 * même contexte mais de seuils différents partageaient une entrée, et le second
 * récupérait silencieusement les résultats filtrés du premier.
 */
function getCacheKey(query: string, config: SmartSearchConfig): string {
	return [
		query,
		config.context || "global",
		config.limit || 20,
		config.minScore ?? 0,
		config.enableSuggestions === false ? "0" : "1",
	].join(":");
}

/**
 * Get cached results if valid
 */
function getCachedResults(query: string, config: SmartSearchConfig): SmartSearchResult[] | null {
	const key = getCacheKey(query, config);
	const entry = searchCache.get(key);

	if (!entry) {
		return null;
	}

	const age = Date.now() - entry.timestamp;
	if (age > CACHE_TTL) {
		searchCache.delete(key);
		return null;
	}

	return entry.results;
}

/**
 * Cache search results
 */
function cacheResults(
	query: string,
	config: SmartSearchConfig,
	results: SmartSearchResult[]
): void {
	const key = getCacheKey(query, config);
	searchCache.set(key, {
		results,
		timestamp: Date.now(),
	});

	// Cleanup old entries (prevent memory leak)
	if (searchCache.size > 100) {
		const now = Date.now();
		for (const [k, entry] of searchCache.entries()) {
			if (now - entry.timestamp > CACHE_TTL) {
				searchCache.delete(k);
			}
		}
	}
}

/**
 * Map search context to prioritized types
 */
const CONTEXT_PRIORITY: Record<SearchContext, SearchableType[]> = {
	aura: ["keshin", "soul", "miximax", "awakening", "modechange", "aura"],
	chara: ["basara", "character"],
	global: [
		"basara",
		"character",
		"skill",
		"keshin",
		"soul",
		"miximax",
		"awakening",
		"modechange",
		"aura",
		"item",
		"tactic",
		"team",
		"article",
		"patchnote",
		"tweet",
		"re",
		"cross",
		"level5",
		"topic",
		"doc",
	],
	item: ["item", "tactic"],
	news: ["article", "patchnote", "tweet", "level5", "topic"],
	passive: ["passive"],
	patchnotes: ["patchnote", "article", "tweet", "topic"],
	skill: ["skill"],
	tactic: ["tactic", "item"],
	team: ["team", "character", "basara"],
};

/**
 * Prioritize results based on search context
 */
function prioritizeResults(
	results: SmartSearchResult[],
	context: SearchContext
): SmartSearchResult[] {
	const priorities = CONTEXT_PRIORITY[context];

	return results.toSorted((a, b) => {
		// First sort by type priority
		const aPriority = priorities.indexOf(a.type);
		const bPriority = priorities.indexOf(b.type);

		if (aPriority !== bPriority) {
			if (aPriority === -1) {
				return 1;
			}
			if (bPriority === -1) {
				return -1;
			}
			return aPriority - bPriority;
		}

		// Then by score
		return b.score - a.score;
	});
}

/**
 * Enhance search results with language detection
 */
function enhanceResults(results: GlobalSearchResult[], query: string): SmartSearchResult[] {
	return results.map((result) => {
		const enhanced: SmartSearchResult = { ...result };

		// Detect matched language for multilingual items
		if (result.data && typeof result.data === "object") {
			const data = result.data as Record<string, unknown>;

			// For BASARA/Characters
			if (data.names && typeof data.names === "object") {
				const names = data.names as Record<string, string>;
				const matchedLang = detectMatchedLanguage(query, names);
				if (matchedLang) {
					enhanced.matchedLanguage = matchedLang;
					// `detectMatchedLanguage` renvoie "JP" mais la clé de l'objet est
					// `ja` : sans ce mapping, un match japonais retombait toujours sur
					// `result.name` (le nom FR) au lieu du nom japonais.
					const langKey = matchedLang === "JP" ? "ja" : (matchedLang.toLowerCase() as "fr" | "en");
					enhanced.matchedName = names[langKey] || result.name;
				}
			}

			// For Skills (uppercase keys from sheet_data)
			if (data.name_FR || data.name_EN || data.name_JA) {
				const matchedLang = detectMatchedLanguage(query, {
					en: data.name_EN as string,
					fr: data.name_FR as string,
					ja: data.name_JA as string,
				});
				if (matchedLang) {
					enhanced.matchedLanguage = matchedLang;
					const langKey = `name_${matchedLang === "JP" ? "JA" : matchedLang}` as
						| "name_FR"
						| "name_EN"
						| "name_JA";
					enhanced.matchedName = (data[langKey] as string) || result.name;
				}
			}

			// For auras, teams, items with lowercase keys (name_fr/name_en/name_ja)
			if (data.name_fr || data.name_en || data.name_ja) {
				const matchedLang = detectMatchedLanguage(query, {
					en: data.name_en as string,
					fr: data.name_fr as string,
					ja: data.name_ja as string,
				});
				if (matchedLang) {
					enhanced.matchedLanguage = matchedLang;
					const langKey = `name_${matchedLang === "JP" ? "ja" : matchedLang.toLowerCase()}`;
					enhanced.matchedName = (data[langKey] as string) || result.name;
				}
			}
		}

		return enhanced;
	});
}

/**
 * Generate "did you mean?" suggestions from all available names
 */
function generateSuggestions(query: string, allResults: GlobalSearchResult[]): string | undefined {
	// Collect all searchable names
	const allNames: string[] = [];

	for (const result of allResults) {
		if (result.data && typeof result.data === "object") {
			const data = result.data as Record<string, unknown>;

			// Noms multilingues — on pousse les valeurs BRUTES : `findClosestMatch`
			// normalise déjà les deux côtés pour le score, et la suggestion est
			// affichée telle quelle à l'utilisateur. Avec `getNameVariations` (qui
			// normalise), la suggestion sortait en minuscules sans accents
			// (« l'etoffe des heros » au lieu de « L'Étoffe des Héros »).
			if (data.names && typeof data.names === "object") {
				for (const value of Object.values(data.names as Record<string, unknown>)) {
					if (typeof value === "string" && value.length > 0) allNames.push(value);
				}
			}
			if (typeof data.name_FR === "string") {
				allNames.push(data.name_FR);
			}
			if (typeof data.name_EN === "string") {
				allNames.push(data.name_EN);
			}
			if (typeof data.name_JA === "string") {
				allNames.push(data.name_JA);
			}
		}
	}

	// Find closest match
	const closest = findClosestMatch(query, allNames, 0.6);
	return closest?.match;
}

/**
 * Smart search with context awareness and language detection
 */
export async function smartSearch(
	query: string,
	rawResults: GlobalSearchResult[],
	config: SmartSearchConfig = {}
): Promise<SmartSearchResult[]> {
	const { context = "global", limit = 20, enableSuggestions = true, minScore = 0 } = config;

	// Check cache first
	const cached = getCachedResults(query, config);
	if (cached) {
		return cached;
	}

	// Enhance results with language detection
	let results = enhanceResults(rawResults, query);

	// Filter by minimum score
	if (minScore > 0) {
		results = results.filter((r) => r.score >= minScore);
	}

	// Prioritize by context
	results = prioritizeResults(results, context);

	// Limit results
	results = results.slice(0, limit);

	// Add suggestions if no results
	if (results.length === 0 && enableSuggestions) {
		const suggestion = generateSuggestions(query, rawResults);
		if (suggestion) {
			// Return empty results with suggestion metadata
			results = [
				{
					data: null,
					id: "",
					name: "",
					score: 0,
					suggestion,
					type: "character",
				} as SmartSearchResult,
			];
		}
	}

	// Cache results
	cacheResults(query, config, results);

	return results;
}

/**
 * Clear search cache (useful for testing or manual cache invalidation)
 */
export function clearSearchCache(): void {
	searchCache.clear();
}

/**
 * Get cache statistics (for debugging)
 */
export function getCacheStats() {
	return {
		entries: Array.from(searchCache.entries()).map(([key, entry]) => ({
			key,
			age: Date.now() - entry.timestamp,
			resultCount: entry.results.length,
		})),
		size: searchCache.size,
	};
}

File: crates/engine/nie-core/src/azalee/smart_search.rs
//! Pure context-aware search from Azalee's `smart-search.ts`.
//!
//! [`smart_search`] only receives an already-built result projection. It
//! enhances multilingual names, filters, prioritizes, limits, and produces a
//! suggestion. It deliberately does not perform HTTP/SQLite/JSON access and
//! does not contain the TypeScript cache, TTL, debounce, or abort-controller
//! behavior. Those are host adapters; the adapter must project its payload to
//! [`GlobalSearchResult`] and can join the returned result back to its payload
//! using `id`.

use std::cmp::Ordering;

use super::search::{
    ClosestMatch, MatchedLanguage, MultilingualNames, detect_matched_language, find_closest_match,
};

/// Searchable IEVR/wiki result categories supported by Azalee.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SearchableType {
    /// BASARA player.
    Basara,
    /// Character.
    Character,
    /// Hissatsu skill.
    Skill,
    /// Item.
    Item,
    /// Tactic.
    Tactic,
    /// Keshin.
    Keshin,
    /// Soul.
    Soul,
    /// Miximax.
    Miximax,
    /// Awakening.
    Awakening,
    /// Mode change.
    ModeChange,
    /// Aura.
    Aura,
    /// Team.
    Team,
    /// News article.
    Article,
    /// Patch note.
    PatchNote,
    /// Passive skill.
    Passive,
    /// X/Twitter post.
    Tweet,
    /// IE RE information.
    Re,
    /// IE Cross information.
    Cross,
    /// LEVEL-5 information.
    Level5,
    /// Topic/documentation information.
    Topic,
    /// Documentation.
    Doc,
}

impl SearchableType {
    /// Returns the stable wire spelling used by the Azalee search payload.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Basara => "basara",
            Self::Character => "character",
            Self::Skill => "skill",
            Self::Item => "item",
            Self::Tactic => "tactic",
            Self::Keshin => "keshin",
            Self::Soul => "soul",
            Self::Miximax => "miximax",
            Self::Awakening => "awakening",
            Self::ModeChange => "modechange",
            Self::Aura => "aura",
            Self::Team => "team",
            Self::Article => "article",
            Self::PatchNote => "patchnote",
            Self::Passive => "passive",
            Self::Tweet => "tweet",
            Self::Re => "re",
            Self::Cross => "cross",
            Self::Level5 => "level5",
            Self::Topic => "topic",
            Self::Doc => "doc",
        }
    }
}

/// Context used to prioritize search result categories.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum SearchContext {
    /// Aura-related entities first.
    Aura,
    /// Characters and BASARA players first.
    Chara,
    /// All categories in the global order.
    #[default]
    Global,
    /// Items and tactics first.
    Item,
    /// News and patch-note categories first.
    News,
    /// Passives only have a preferred category.
    Passive,
    /// Patch notes and related news first.
    PatchNotes,
    /// Skills first.
    Skill,
    /// Tactics and items first.
    Tactic,
    /// Teams, characters, and BASARA players first.
    Team,
}

/// Owned multilingual names accepted by the pure search projection.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct SearchNames {
    /// French name.
    pub french: Option<String>,
    /// English name.
    pub english: Option<String>,
    /// Japanese name.
    pub japanese: Option<String>,
}

impl SearchNames {
    /// Builds names from the language values supplied by a host adapter.
    #[must_use]
    pub fn new(french: Option<&str>, english: Option<&str>, japanese: Option<&str>) -> Self {
        Self {
            french: french.map(str::to_owned),
            english: english.map(str::to_owned),
            japanese: japanese.map(str::to_owned),
        }
    }

    fn has_value(&self) -> bool {
        self.french
            .as_deref()
            .is_some_and(|value| !value.is_empty())
            || self
                .english
                .as_deref()
                .is_some_and(|value| !value.is_empty())
            || self
                .japanese
                .as_deref()
                .is_some_and(|value| !value.is_empty())
    }

    fn borrowed(&self) -> MultilingualNames<'_> {
        MultilingualNames {
            french: self.french.as_deref(),
            english: self.english.as_deref(),
            japanese: self.japanese.as_deref(),
        }
    }

    fn value_for(&self, language: MatchedLanguage) -> Option<&str> {
        match language {
            MatchedLanguage::French => self.french.as_deref(),
            MatchedLanguage::English => self.english.as_deref(),
            MatchedLanguage::Japanese => self.japanese.as_deref(),
        }
    }

    fn values<'a>(&'a self, output: &mut Vec<&'a str>) {
        for value in [&self.french, &self.english, &self.japanese] {
            if let Some(value) = value.as_deref().filter(|value| !value.is_empty()) {
                output.push(value);
            }
        }
    }
}

/// Name fields recognized from Azalee's JSON result shapes.
///
/// `names` corresponds to character/BASARA rows, `skill_names` to the
/// uppercase `name_FR`/`name_EN`/`name_JA` fields, and `localized_names` to
/// lowercase `name_fr`/`name_en`/`name_ja` fields. The latter is used for
/// annotation but, like the TypeScript source, is not used for suggestions.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct SearchData {
    /// Character/BASARA `names` object.
    pub names: Option<SearchNames>,
    /// Skill sheet `name_FR`/`name_EN`/`name_JA` fields.
    pub skill_names: Option<SearchNames>,
    /// Lowercase `name_fr`/`name_en`/`name_ja` fields.
    pub localized_names: Option<SearchNames>,
}

/// A result projection supplied by a database or web adapter.
#[derive(Debug, Clone, PartialEq)]
pub struct GlobalSearchResult {
    /// Stable result identifier.
    pub id: String,
    /// Default display name.
    pub name: String,
    /// Result category.
    pub result_type: SearchableType,
    /// Upstream relevance score.
    pub score: f32,
    /// Recognized multilingual name fields.
    pub data: Option<SearchData>,
}

/// Search result after language annotation and optional suggestion metadata.
#[derive(Debug, Clone, PartialEq)]
pub struct SmartSearchResult {
    /// Stable result identifier.
    pub id: String,
    /// Default display name.
    pub name: String,
    /// Result category.
    pub result_type: SearchableType,
    /// Upstream relevance score.
    pub score: f32,
    /// Recognized multilingual name fields.
    pub data: Option<SearchData>,
    /// Language that matched the query.
    pub matched_language: Option<MatchedLanguage>,
    /// Original spelling of the matched name.
    pub matched_name: Option<String>,
    /// Original spelling of a closest suggestion when no result survives.
    pub suggestion: Option<String>,
}

/// Configuration for the pure smart-search operation.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct SmartSearchConfig {
    /// Context-specific result priority.
    pub context: SearchContext,
    /// Maximum number of returned results.
    pub limit: usize,
    /// Whether to emit a suggestion placeholder when no result survives.
    pub enable_suggestions: bool,
    /// Minimum upstream score. Values `<= 0` disable filtering.
    pub min_score: f32,
}

impl Default for SmartSearchConfig {
    fn default() -> Self {
        Self {
            context: SearchContext::Global,
            limit: 20,
            enable_suggestions: true,
            min_score: 0.0,
        }
    }
}

/// Applies Azalee's context-aware enhancement, ordering, filtering, limiting,
/// and suggestion rules to already-loaded results.
#[must_use]
pub fn smart_search(
    query: &str,
    raw_results: &[GlobalSearchResult],
    config: SmartSearchConfig,
) -> Vec<SmartSearchResult> {
    let mut results: Vec<_> = raw_results
        .iter()
        .map(|result| enhance(result, query))
        .collect();

    if config.min_score > 0.0 {
        results.retain(|result| result.score >= config.min_score);
    }

    let priorities = priorities(config.context);
    results.sort_by(|left, right| {
        type_priority(left.result_type, priorities)
            .cmp(&type_priority(right.result_type, priorities))
            .then_with(|| {
                right
                    .score
                    .partial_cmp(&left.score)
                    .unwrap_or(Ordering::Equal)
            })
    });
    results.truncate(config.limit);

    if results.is_empty()
        && config.enable_suggestions
        && let Some(suggestion) = generate_suggestion(query, raw_results)
    {
        results.push(SmartSearchResult {
            id: String::new(),
            name: String::new(),
            result_type: SearchableType::Character,
            score: 0.0,
            data: None,
            matched_language: None,
            matched_name: None,
            suggestion: Some(suggestion.value),
        });
    }

    results
}

fn enhance(result: &GlobalSearchResult, query: &str) -> SmartSearchResult {
    let mut enhanced = SmartSearchResult {
        id: result.id.clone(),
        name: result.name.clone(),
        result_type: result.result_type,
        score: result.score,
        data: result.data.clone(),
        matched_language: None,
        matched_name: None,
        suggestion: None,
    };

    if let Some(data) = result.data.as_ref() {
        for names in [&data.names, &data.skill_names, &data.localized_names]
            .into_iter()
            .flatten()
            .filter(|names| names.has_value())
        {
            if let Some(language) = detect_matched_language(query, names.borrowed()) {
                enhanced.matched_language = Some(language);
                enhanced.matched_name = Some(
                    names
                        .value_for(language)
                        .filter(|value| !value.is_empty())
                        .unwrap_or(&result.name)
                        .to_owned(),
                );
            }
        }
    }

    enhanced
}

fn generate_suggestion(query: &str, raw_results: &[GlobalSearchResult]) -> Option<ClosestMatch> {
    let mut names = Vec::new();
    for result in raw_results {
        if let Some(data) = result.data.as_ref() {
            if let Some(values) = data.names.as_ref() {
                values.values(&mut names);
            }
            if let Some(values) = data.skill_names.as_ref() {
                values.values(&mut names);
            }
        }
    }
    find_closest_match(query, &names, 0.6)
}

fn priorities(context: SearchContext) -> &'static [SearchableType] {
    match context {
        SearchContext::Aura => &[
            SearchableType::Keshin,
            SearchableType::Soul,
            SearchableType::Miximax,
            SearchableType::Awakening,
            SearchableType::ModeChange,
            SearchableType::Aura,
        ],
        SearchContext::Chara => &[SearchableType::Basara, SearchableType::Character],
        SearchContext::Global => &[
            SearchableType::Basara,
            SearchableType::Character,
            SearchableType::Skill,
            SearchableType::Keshin,
            SearchableType::Soul,
            SearchableType::Miximax,
            SearchableType::Awakening,
            SearchableType::ModeChange,
            SearchableType::Aura,
            SearchableType::Item,
            SearchableType::Tactic,
            SearchableType::Team,
            SearchableType::Article,
            SearchableType::PatchNote,
            SearchableType::Tweet,
            SearchableType::Re,
            SearchableType::Cross,
            SearchableType::Level5,
            SearchableType::Topic,
            SearchableType::Doc,
        ],
        SearchContext::Item => &[SearchableType::Item, SearchableType::Tactic],
        SearchContext::News => &[
            SearchableType::Article,
            SearchableType::PatchNote,
            SearchableType::Tweet,
            SearchableType::Level5,
            SearchableType::Topic,
        ],
        SearchContext::Passive => &[SearchableType::Passive],
        SearchContext::PatchNotes => &[
            SearchableType::PatchNote,
            SearchableType::Article,
            SearchableType::Tweet,
            SearchableType::Topic,
        ],
        SearchContext::Skill => &[SearchableType::Skill],
        SearchContext::Tactic => &[SearchableType::Tactic, SearchableType::Item],
        SearchContext::Team => &[
            SearchableType::Team,
            SearchableType::Character,
            SearchableType::Basara,
        ],
    }
}

fn type_priority(result_type: SearchableType, priorities: &[SearchableType]) -> usize {
    priorities
        .iter()
        .position(|candidate| *candidate == result_type)
        .unwrap_or(usize::MAX)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn result(
        id: &str,
        name: &str,
        result_type: SearchableType,
        score: f32,
        data: SearchData,
    ) -> GlobalSearchResult {
        GlobalSearchResult {
            id: id.to_owned(),
            name: name.to_owned(),
            result_type,
            score,
            data: Some(data),
        }
    }

    fn character_data() -> SearchData {
        SearchData {
            names: Some(SearchNames::new(
                Some("Mark Evans"),
                Some("Mark Evans"),
                Some("円堂 守"),
            )),
            ..Default::default()
        }
    }

    #[test]
    fn prioritizes_context_then_score_and_respects_limit() {
        let raw = vec![
            result(
                "team",
                "Mark's team",
                SearchableType::Team,
                1.0,
                character_data(),
            ),
            result(
                "skill",
                "Mark skill",
                SearchableType::Skill,
                0.1,
                character_data(),
            ),
            result(
                "chara",
                "Mark",
                SearchableType::Character,
                0.8,
                character_data(),
            ),
        ];
        let results = smart_search(
            "Mark",
            &raw,
            SmartSearchConfig {
                context: SearchContext::Skill,
                limit: 2,
                ..Default::default()
            },
        );
        assert_eq!(results.len(), 2);
        assert_eq!(results[0].result_type, SearchableType::Skill);
        assert_eq!(results[1].result_type, SearchableType::Team);
    }

    #[test]
    fn enhances_uppercase_and_lowercase_name_shapes() {
        let raw = vec![result(
            "skill",
            "Tornade de feu",
            SearchableType::Skill,
            1.0,
            SearchData {
                skill_names: Some(SearchNames::new(
                    Some("Tornade de feu"),
                    Some("Fire Tornado"),
                    Some("ファイアトルネード"),
                )),
                ..Default::default()
            },
        )];
        let results = smart_search("ファイア", &raw, SmartSearchConfig::default());
        assert_eq!(results[0].matched_language, Some(MatchedLanguage::Japanese));
        assert_eq!(
            results[0].matched_name.as_deref(),
            Some("ファイアトルネード")
        );

        let localized = vec![result(
            "item",
            "Objet",
            SearchableType::Item,
            1.0,
            SearchData {
                localized_names: Some(SearchNames::new(Some("Objet"), Some("Item"), None)),
                ..Default::default()
            },
        )];
        let localized_results = smart_search("Item", &localized, SmartSearchConfig::default());
        assert_eq!(
            localized_results[0].matched_language,
            Some(MatchedLanguage::English)
        );
        assert_eq!(localized_results[0].matched_name.as_deref(), Some("Item"));
    }

    #[test]
    fn filters_before_suggesting_and_keeps_raw_spelling() {
        let raw = vec![result(
            "team",
            "L'Étoffe des Héros",
            SearchableType::Team,
            1.0,
            SearchData {
                names: Some(SearchNames::new(
                    Some("L'Étoffe des Héros"),
                    Some("Legendary Heroes"),
                    Some("英雄の証"),
                )),
                ..Default::default()
            },
        )];
        let results = smart_search(
            "L'Étoffe des Hero",
            &raw,
            SmartSearchConfig {
                min_score: 2.0,
                ..Default::default()
            },
        );
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].suggestion, Some("L'Étoffe des Héros".to_owned()));
    }

    #[test]
    fn suggestion_uses_only_the_source_name_fields() {
        let raw = vec![result(
            "item",
            "Objet",
            SearchableType::Item,
            1.0,
            SearchData {
                localized_names: Some(SearchNames::new(Some("Objet réel"), None, None)),
                ..Default::default()
            },
        )];
        let results = smart_search(
            "Objet reel",
            &raw,
            SmartSearchConfig {
                min_score: 2.0,
                ..Default::default()
            },
        );
        assert!(results.is_empty());
    }
}
